keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md

```diff
@@ -0,0 +1,430 @@
+domains.
+
+## Motivation
+
+There are scenarios where a ReplicaSet can reach an imbalanced state across
+failure domains. See [user stories](#user-stories) to see one such scenario
+and how a randomized approach solves the issue.
+
+### Goals
+
+- A randomized algorithm for Pod selection when downscaling ReplicaSets.
+- Softly honor the heuristic that prefers to downscale newer Pods first.
+- Validate that the approach is able to get a ReplicaSet out of an imbalanced
+  state.
+
+### Non-Goals
+
+- Provide guarantees of preserving balance during scale down. This introduces
+  a violation of separation of concerns between the ReplicaSet controller and
+  kube-scheduler.
+- Preserve the existing behavior that always downscales the newer Pods first.
+  This order was never guaranteed in the API or user documentation.
+
+## Proposal
+
+
+### User Stories
+
+#### Story 1
+
+This story shows an imbalance cycle after a failure domain fails or gets
+upgraded.
+
+1. Assume a ReplicaSet has 2N pods evenly distributed across 2 failure domains,
+   thus each has N pods.
+2. An upgrade happens adding a new available domain and the ReplicaSet is upscaled
+   to 3N. The new domain now holds all the youngest pods due to scheduler spreading.
+3. ReplicaSet is downscaled to 2N again. Due to the downscaling preference, all
+   the Pods from one domain are removed, leading to imbalance. The situation
+   doesn't improve with repeated upscale and downscale steps. Instead, a
+   randomized approach leaves about 2/3*N nodes in each
+   failure domain.
+
+### Notes/Constraints/Caveats
+
+The original heuristic, dowscaling the youngest Pods first, has its benefits.
+Newer Pods might not have finished starting up (or warming up) and are likely
+to have less active connections than older Pods. However, this distinction
+doesn't generally apply once Pods have been running steadily for some time.
+
+A purely randomized approach would break those assumptions, potentially
+leading to services disruption. Choosing a heuristic could be left to the user.
+On the other hand, certain workloads take a long time to warm up and, at the
+same time, require high availability.
+
+### Risks and Mitigations
+
+Certain users might be relaying in the existing downscaling heuristic. However,
+there are a number of reasons why we don't need to preserve such behavior as is:
+
+- The behavior is not documented in the API or user docs, only in code.
+- The heuristic is applied last, after other criteria have been applied. In
+  particular, the heuristic doesn't apply when multiple Pods are in the same
+  Node or if a particular Pod has had several container restarts. While users
+  can enforce one Pod per Node with certain scheduling features, there is no
+  workaround for other criteria.
+- We are not proposing to entirely remove the heuristic, just make it more
+  laxed.
+- We are introducing a [related feature](git.k8s.io/enhancements/keps/sig-apps/1828-delete-priority-annotations)
```

Comment:
nit: link not working

Reply:
Updated, seems that kep number/name changed